Add validation for VK_KHR_fragment_shading_rate

Tobski opened the pull request.

Tobski commented:
This is missing a handful of VUs which require GPU validation support, which I'll attempt to get into a future PR once this has landed.

sfricke-samsung commented on layers/core_validation.cpp:
```diff
@@ -1871,6 +1877,104 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
+
+            if (pFragmentShadingRateState->fragmentSize.width == 0) {
+                skip |= LogError(device, "VUID-VkGraphicsPipelineCreateInfo-pDynamicState-04494",
+                                 "Fragment width of %u has been specified in %s.", pFragmentShadingRateState->fragmentSize.width,
```
I have been on an effort recently to add the function name to each message as its very helpful for developers, so could you please add `vkCreateGraphicsPipelines(): ` before all of these... realize the others might not all have been fixed yet (same with all newly added `LogError` also)

sfricke-samsung commented on layers/shader_validation.cpp:
```diff
@@ -3458,6 +3460,72 @@ bool CoreChecks::ValidatePointListShaderState(const PIPELINE_STATE *pipeline, SH
+            if (insn.word(2) == spv::DecorationBuiltIn) {
+                if (insn.word(3) == spv::BuiltInPrimitiveShadingRateKHR) {
+                    primitiverate_written = IsBuiltInWritten(src, insn, entrypoint);
+                } else if (insn.word(4) == spv::BuiltInViewportIndex) {
```
shouldn't these all be `insn.word(3)`?

Tobski replied:
Oops, copy+paste error, thanks.

Tobski replied:
(Noting that actually testing this properly atm has been impossible because I don't currently have a driver that can create a shader module from appropriate code...)

SaschaWillems commented:
I'm currently working on a sample using this extension, and was getting crashes when using this branch. It looks like the resolve attachment check wrongly uses the color attachment count for it's loop, which results in a crash if you don't use any resolve attachments:

![image](https://user-images.githubusercontent.com/10283127/97741207-afd78400-1ae2-11eb-93dc-65a19336a9c4.png)

SaschaWillems commented:
Also judging by the spec:
![image](https://user-images.githubusercontent.com/10283127/97744758-14e1a880-1ae8-11eb-8e0e-8c6a48bc0400.png)

It looks like I don't have to specify `VkPipelineFragmentShadingRateStateCreateInfoKHR` at pipeline creation, esp. as I'm using only a sample rate attachment.

But if I don't specify that structure at pipeline creation time, validation will crash:

![image](https://user-images.githubusercontent.com/10283127/97744919-583c1700-1ae8-11eb-9d8f-0e9cd631fceb.png)

Shouldn't that code check if the application is actually using

SaschaWillems commented:
I also noticed odd behavior with `vkGetPhysicalDeviceFragmentShadingRatesKHR`. Calling this for getting the number of shading rates with the validation layers enabled, it doesn't update `pFragmentShadingRateCount`. So if I initialize that to zero, it stays zero after calling that function. But only when validation is enabled. If I disable validation, the function return the correct number of fragment shading rates available for my device. Not sure why the validation layers result in this behavior.

Tobski commented:
@SaschaWillems thanks for the feedback! I'll take a look and maybe add some positive tests in to cover this.

Tobski commented:
> It looks like this is lacking a check for if (subpassInfo.pResolveAttachments).

Whoops yep that's an easy fix.

> I'm not sure if it's the right way to just check if the extension is present and then run this check. What if I have mixed pipelines, and not all of them make use of fragment shading rate? This will crash for all of those, unless I set that part of the pipeline creation info. At least that code should check if pFragmentShadingRateState is not null.

Yea that should just check to make sure the structure is there (since the default values are definitely ok in all cases) - I'll add that check.

> I also noticed odd behavior with vkGetPhysicalDeviceFragmentShadingRatesKHR. Calling this for getting the number of shading rates with the validation layers enabled, it doesn't update pFragmentShadingRateCount. So if I initialize that to zero, it stays zero after calling that function. But only when validation is enabled. If I disable validation, the function return the correct number of fragment shading rates available for my device. Not sure why the validation layers result in this behavior.

Hum, that is a weird one. I'm no sure why it'd be doing that - it should just be doing pass through. I haven't been able to test this locally since I don't currently have a GPU/driver combo that supports this extension... I'll see if I can figure it out from code inspection at least.

Tony-LunarG commented:
I'm seeing a problem with one of the new tests - we might need a new glslang for this PR

[ RUN      ] VkLayerTest.InvalidPrimitiveFragmentShadingRateWriteMultiViewportLimit
WARNING: 0:2: '#extension' : extension not supported: GL_EXT_fragment_shading_rate
ERROR: 0:4: 'gl_PrimitiveShadingRateEXT' : undeclared identifier
ERROR: 0:4: '' : compilation terminated
ERROR: 2 compilation errors.  No code generated.

C:\Users\tony\Vulkan-ValidationLayers\tests\vktestbinding.cpp(730): error: Failed
Assertion: `vk::CreateShaderModule(dev.handle(), &info, 0, &handle) == VK_SUCCESS'
Unexpected: Validation Error: [ UNASSIGNED-GeneralParameterError-RequiredParameter ] Object 0: handle = 0x1f287e267f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x99fb7dfd | vkCreateGraphicsPipelines: required parameter pCreateInfos[0].pStages[0].module specified as VK_NULL_HANDLE
Unexpected: Validation Error: [ VUID-VkPipelineShaderStageCreateInfo-module-parameter ] Object 0: handle = 0x1f2fcf16020, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0xfd71dc70 | Invalid VkShaderModule Object 0x0. The Vulkan spec states: module must be a valid VkShaderModule handle (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkPipelineShaderStageCreateInfo-module-parameter)
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.

Tobski commented:
@SaschaWillems ok all of the issues you pointed out should be fixed in the latest version, though I'm still waiting to see what CI does before this is good to go.

As for this:

> I also noticed odd behavior with vkGetPhysicalDeviceFragmentShadingRatesKHR. Calling this for getting the number of shading rates with the validation layers enabled, it doesn't update pFragmentShadingRateCount. So if I initialize that to zero, it stays zero after calling that function. But only when validation is enabled. If I disable validation, the function return the correct number of fragment shading rates available for my device. Not sure why the validation layers result in this behavior.

So the reason this is breaking is because that function is queried via vkGetInstanceProcAddr(), which means it comes from the loader, so you need an updated loader for this function to work... :/

SaschaWillems commented:
> So I think the reason this is breaking is because that function is queried via vkGetInstanceProcAddr(), which means it comes from the loader, so you need an updated loader for this function to work... :/

Thanks for the hint. That makes sense. For the Khronos Vulkan Samples, we use Volk as a loader, which already is updated to the latest header.

Is there a way to change the loader path for the layers, so I can compile a recent loader and use that instead?

mark-lunarg commented:
@SaschaWillems, you can clone the `Vulkan-Loader` repo and follow the directions to run update_deps.py and build it yourself.

To use the new loader, copy the DLL to your application's working directory or replace the one in `windows/system32`.
On Linux, you can point to a new loader using `LD_LIBRARY_PATH`.

SaschaWillems commented:
Thanks. That did work for me. With the replaced loader, `vkGetPhysicalDeviceFragmentShadingRatesKHR` returns proper values with validation enabled. With the two fixes mentioned before, validation now works as expected for the sample I'm currently working on.

SaschaWillems commented:
I'm not sure on parameter validation for `vkGetPhysicalDeviceFragmentShadingRatesKHR` though. To get the available shading rates, you call the function twice (similar to many other vk getters):

	vkGetPhysicalDeviceFragmentShadingRatesKHR(get_device().get_gpu().get_handle(), &fragment_shading_rate_count, nullptr);
	if (fragment_shading_rate_count > 0)
	{
		fragment_shading_rates.resize(fragment_shading_rate_count);
		vkGetPhysicalDeviceFragmentShadingRatesKHR(get_device().get_gpu().get_handle(), &fragment_shading_rate_count, fragment_shading_rates.data());
	}

The second call will trigger parameter validation for all entries in the vector for sType not being `VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FRAGMENT_SHADING_RATE_KHR`. 

Not sure how this is handled in other similar validations, but it feels kinda odd having to prefill all vector entries with an sType, even though I'm just fetching them via the vkGet function.

Tobski commented:
> Not sure how this is handled in other similar validations, but it feels kinda odd having to prefill all vector entries with an sType, even though I'm just fetching them via the vkGet function

This is apparently consistent, the same is true for physical device properties/features, it's just that usually you don't need to pre-fill an entire array...

Tobski commented:
> I'm seeing a problem with one of the new tests - we might need a new glslang for this PR

Ah whoops I missed that https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2258 didn't upgrade the glslang version. I'll attempt to sort this out.

Tobski commented:
@mark-lunarg all of the issues should be resolved now, and CI is passing, so this should be ready for final review...

sfricke-samsung commented on layers/buffer_validation.cpp:
```diff
@@ -5155,20 +5163,31 @@ bool CoreChecks::PreCallValidateCreateImageView(VkDevice device, const VkImageVi
+        if (device_extensions.vk_nv_shading_rate_image && enabled_features.shading_rate_image.shadingRateImage) {
+            if (image_usage & VK_IMAGE_USAGE_SHADING_RATE_IMAGE_BIT_NV) {
+                if (view_type != VK_IMAGE_VIEW_TYPE_2D && view_type != VK_IMAGE_VIEW_TYPE_2D_ARRAY) {
+                    skip |= LogError(pCreateInfo->image, "VUID-VkImageViewCreateInfo-image-02086",
```
> VUID-VkImageViewCreateInfo-image-02086
If image was created with usage containing VK_IMAGE_USAGE_FRAGMENT_SHADING_RATE_ATTACHMENT_BIT_KHR, viewType must be VK_IMAGE_VIEW_TYPE_2D or VK_IMAGE_VIEW_TYPE_2D_ARRAY

So I was curious why you needed to wrap this all in

`if (device_extensions.vk_nv_shading_rate_image && enabled_features.shading_rate_image.shadingRateImage)` as usually these just work out cause they look for the extension enums... Like checking for `VK_IMAGE_USAGE_SHADING_RATE_IMAGE_BIT_NV` should be enough, no need to wrap with an extra `if`

Also `02086` seems wrong here, not sure if this VU got changed without the number updating?

Tobski replied:
Hmm, might be that the VUID got shuffled down one when I wasn't looking or something - 2087 should be checked explicitly for the NV shading rate extension but not the KHR fragment shading rate extension. I'll take a look.

Tobski replied:
Ahhh I remember what happened here - I had planned to put down a different output depending on the extension - NV usage bit if the NV extension was enabled, KHR otherwise. But given I haven't done this for literally anything else I'm going to tease this out.

sfricke-samsung commented on layers/buffer_validation.cpp:
```diff
@@ -5155,20 +5163,31 @@ bool CoreChecks::PreCallValidateCreateImageView(VkDevice device, const VkImageVi
             }
         }
 
+        if (device_extensions.vk_khr_fragment_shading_rate &&
```
again, not sure why need teh `device_extensions.vk_khr_fragment_shading_rate`  check here

Tobski replied:
The limit being checked here doesn't apply to the NV extension, which uses the same usage flag. So have to check that the fsr extension is in use. I should probably check the feature instead though since that's the thing that gates it, not the extension.

Tobski replied:
I'll file an issue to fix the VU in the spec because this isn't obvious from the built spec - it's only apparent in the source. Ditto for 02087...

sfricke-samsung commented on layers/core_validation.cpp:
```diff
@@ -1871,6 +1877,104 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
                          "Extended dynamic state used by the extendedDynamicState feature is not enabled");
     }
 
+    if (device_extensions.vk_khr_fragment_shading_rate) {
```
why not just make this

```
const VkPipelineFragmentShadingRateStateCreateInfoKHR *pFragmentShadingRateState =
            lvl_find_in_chain<VkPipelineFragmentShadingRateStateCreateInfoKHR>(pPipeline->graphicsPipelineCI.pNext);

if (pFragmentShadingRateState) {
```

Tobski replied:
I always forget the rules of what happens if an extension isn't enabled 🤔 I'll remove the check though, I think it's probably unnecessary.

sfricke-samsung replied:
There should be auto-generation for use of functions without the extensions enabled, but I guess its more of a "it will be validated somewhere else" and prevents from having to `if (ext.enabled)` check anywhere, to my knowledge

SaschaWillems commented:
I'm getting a SPIR-V invalid capability validation message when using `gl_ShadingRateEXT` in a fragment shader:

```bash
[error] [framework\core\instance.cpp:41] 7060244 - UNASSIGNED-CoreValidation-Shader-InconsistentSpirv: Validation Error: [ UNASSIGNED-CoreValidation-Shader-InconsistentSpirv ] Object 0: handle = 0x20d1b56f068, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x6bbb14 | SPIR-V module not valid: Invalid capability operand: 4422
```

Tobski commented:
> ```shell
> UNASSIGNED-CoreValidation-Shader-InconsistentSpirv
> ```

@SaschaWillems are you definitely using up to date spirv tools in your build? I don't seem to get that error when the capability is declared. If you're sure you're up to date, is the shader somewhere I can look at to help debug it, or could you post it here?

SaschaWillems commented:
We are using glslang to compile to SPIR-V at runtime in the Khronos Samples repo. I have just updated the glslang submodule to the latest commit, but I'm still getting that validation error.

Tobski commented:
> 
> 
> We are using glslang to compile to SPIR-V at runtime in the Khronos Samples repo. I have just updated the glslang submodule to the latest commit, but I'm still getting that validation error.

Ok sorry it took a while to get back to this - so I pulled a clean copy of the latest glslang version, pulled the known-good spirv-tools associated with it (which is the same as this PR), and ran [your fragment shader](https://github.com/SaschaWillems/Vulkan-Samples/blob/vk_khr_fragment_shading_rate/shaders/fragment_shading_rate/scene.frag) through glslang - and got no validation errors - spirv-dis from the same repository also correctly outputs the right thing. Is it possible you've got an out of date spirv-tools with the version of glslang you're building?

@Tony-LunarG you seemed to be running these tests at some point - are you seeing anything like this?

Tony-LunarG commented:
I don't see any inconsistent spirv errors running the tests.  I do see a crash:

[ RUN      ] VkLayerTest.InvalidFragmentShadingRatePipeline
[       OK ] VkLayerTest.InvalidFragmentShadingRatePipeline (1291 ms)
[ RUN      ] VkLayerTest.InvalidFragmentShadingRatePipelineFeatureUsage
[       OK ] VkLayerTest.InvalidFragmentShadingRatePipelineFeatureUsage (1200 ms)
[ RUN      ] VkLayerTest.InvalidFragmentShadingRatePipelineCombinerOpsLimit
             TEST SKIPPED: requires fragmentShadingRateNonTrivialCombinerOps to be unsupported.
[       OK ] VkLayerTest.InvalidFragmentShadingRatePipelineCombinerOpsLimit (148 ms)
[ RUN      ] VkLayerTest.InvalidPrimitiveFragmentShadingRateWriteMultiViewportLimit
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.

Tobski commented:
@Tony-LunarG Hm. I can't reproduce that here but I don't have a GPU with an implementation currently... if you're able to get any additional debug info out (e.g. backtrace) that would be really helpful. I'll try to do some investigation on our end.

Tony-LunarG commented:
shader_validation.cpp 3541: pipeline->graphicsPipelineCI.pViewportState->viewportCount != 1 && primitiverate_written) {
but pipeline->graphicsPipelineCI.pViewportState is NULL because it was never copied into the safe struct because pipeline->graphicsPipelineCI.pRasterizationState->rasterizationDiscardEnable is true. I'm also not seeing 04503 even without discard because pipeline->graphicsPipelineCI.pViewportState->viewportCount == 1

Tobski commented:
> 
> 
> shader_validation.cpp 3541: pipeline->graphicsPipelineCI.pViewportState->viewportCount != 1 && primitiverate_written) {
> but pipeline->graphicsPipelineCI.pViewportState is NULL because it was never copied into the safe struct because pipeline->graphicsPipelineCI.pRasterizationState->rasterizationDiscardEnable is true. I'm also not seeing 04503 even without discard because pipeline->graphicsPipelineCI.pViewportState->viewportCount == 1

Ah amazing - thanks Tony, was starting to tear my hair out debugging this blind 😬. Well at least it's a test issue instead of a layer issue, should be easy to fix.

Tobski commented:
@Tony-LunarG I pushed an update addressing those issues; it _looks_ like it's working now, but I don't know for sure. I'll try to get someone over here with a board to run these tests, but if you are able to run them in the meantime and verify whether the fix worked or not that would be awesome 👍

Tony-LunarG commented:
I get the same NULL dereference in the same place.

Tobski commented:
This is what I get for trying to push something right as I sign off for the evening - I didn't actually push the changes, just did an empty force-push 🤦 

Actually pushed my changes this time...

Tony-LunarG commented:
No crashes, 5 failures

[ShadingRateLog.txt](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/5576397/ShadingRateLog.txt)

Tobski commented:
@Tony-LunarG thanks for all your help - I've managed to reproduce those issues locally with some hacking, and AFAICT I've fixed all of them in the latest change. I've managed to get hold of a remote machine I can verify on either later today or tomorrow, but if you are able to test before then that would be helpful. Given you're on an NV card you'll be able to double check some tests that I won't be able to anyway.

Tobski commented:
@Tony-LunarG I got everything working on my end AFAICT; if you could confirm for any tests that require NV_shading_rate that would be great 👍 - then hopefully we can land this.

Tony-LunarG commented:
@Tobski One left.  I'm OOO the rest of this week
[ShadingRateLog2.txt](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/5593231/ShadingRateLog2.txt)

Tobski commented:
@Tony-LunarG ok that was just a silly mistake in the VIL test as far as I can tell (which I had absolutely no way to test) - should be fixed now. Will watch CI to make sure it's passing but then hopefully we can get this in...

Tony-LunarG commented:
All tests pass

mark-lunarg commented:
LGTM.  Rebased and cleaned up variable names.  Running through CI one final time and will push when it's clean.  Thanks!

mark-lunarg commented:
Passed CI, rebased and pushed to Master.

Tobski commented:
Thanks!
